415 feat create people overview by faculty endpoint   be

SirAntonySir opened the pull request:
Adding people crawler and people endpoint in BE. 

Adjusting faculty to contain id and code in model.

SirAntonySir committed:
initialize the people by faculty endpoint

SirAntonySir committed:
- Introduced new data models for Person, PersonRole, PersonCourse, and PersonBasicInfo to encapsulate detailed information about individuals.
- Added endpoints for retrieving people lists, filtering by faculty, and fetching individual person details.
- Implemented a data collection service to gather and store people data from the LSF system, including roles and courses.
- Enhanced the database schema with new tables for storing people, their roles, and courses, ensuring a structured approach to data management.

SirAntonySir committed:
- Enhanced the PeopleService to collect and store people data role by role, improving data processing efficiency.
- Introduced crawl_single_role method in LSFPersonCrawler to fetch individuals associated with a specific role ID, including detailed logging for better tracking.
- Updated person ID generation to utilize personal.pid from profile URLs for improved uniqueness.
- Refactored upsert methods to prevent duplication of roles and courses in the database.

SirAntonySir committed:
Refactor LSFPersonCrawler and PeopleService to enhance deduplication and error handling

- Set REQUEST_DELAY to 0 in LSFPersonCrawler to optimize request timing.
- Implemented deduplication for courses and processed people in LSFPersonCrawler to prevent duplicate entries.
- Improved error handling in PeopleService for database operations, specifically for duplicate key violations, ensuring smoother processing of person data.
- Enhanced role and course ID generation to include person-specific identifiers, reducing collision risks.

SirAntonySir committed:
Enhance People API to support faculty ID filtering

- Added `faculty_id` field to `Person` and `PersonSummary` models for better identification.
- Updated `get_people` endpoint to filter by `faculty_id` instead of faculty enum.
- Introduced a new endpoint `get_people_by_faculty_id` for retrieving people by specific faculty ID.
- Enhanced `PeopleService` to map faculty ID to German faculty names and retrieve associated IDs from the database.
- Refactored existing faculty filtering logic to accommodate both faculty enum and ID.

SirAntonySir committed:
Refactor People API endpoints for improved pagination and faculty filtering

- Updated `get_people` endpoint to apply pagination only when no faculty filter is provided.
- Enhanced `get_people_by_faculty_enum` to disable pagination, returning all results for a specific faculty.
- Introduced `get_faculty_enum_by_id` method in `PeopleService` to retrieve faculty enums based on faculty ID.
- Improved error handling and query logic in `PeopleService` for better performance and clarity.

SirAntonySir committed:
- Removed `faculty` and `faculty_id` fields from `Person` and `PersonSummary` models, replacing them with a more streamlined approach using `faculty_enum`.
- Updated `get_people_by_faculty_enum` endpoint to accept `faculty_code` instead of `faculty` enum, enhancing clarity and usability.
- Introduced a new endpoint `get_people_by_faculty_id` for retrieving people based on faculty ID.
- Refactored `PeopleService` methods to utilize faculty codes, improving database queries and error handling.
- Adjusted the `FacultyTable` and `FacultyTranslationTable` to use integer IDs instead of strings for better consistency.

SirAntonySir committed:
delete database services

rpgraffi commented on api/src/v1/people/routers/people_router.py:
lazy scroll would be nice for this. Im not sure if that works with pagination implementation.

rpgraffi replied:
is lazy scroll planned for the frontend, if yes check with ux how the concept of loading states is

rpgraffi commented on api/src/v1/people/services/people_service.py:
im not sure if we need the faculty name here. its should be already loaded in the frontend. btw it should be called from the cms. we store the factulties there. 

generally i think we don't need the facultiy name. 
only thing we need the faculty for is to process the request from the frontend to return people by faculty.

rpgraffi commented on data_fetcher/src/people/services/people_service.py:
```diff
@@ -0,0 +1,322 @@
+                    # Re-raise other types of errors
+                    raise
+
+    def _generate_person_id(self, person_data: dict) -> str:
```
is there a way to create an id always the same way? so we dont have duplicate contacts when some data changes.

rpgraffi commented on shared/src/enums/people_enums.py:
maybe we should save the translations in the database - if thats not the case yet

SirAntonySir committed:
Refactor People API and Data Fetcher for improved structure and functionality

- Consolidated person-related models by importing from a shared module, enhancing code maintainability.
- Updated `people_router` to utilize a new `PeopleAPIService`, streamlining service interactions.
- Enhanced error handling in faculty filtering logic within the `get_people` endpoint.
- Replaced the deprecated `PeopleCollector` with `PeopleCollectorV2` in the data fetcher, improving data collection efficiency.
- Introduced new logging capabilities for better tracking of data fetching processes.
- Refactored enums for academic titles and roles to improve clarity and consistency across the application.
